File: licensing/sdk/python/ras_license_client.py
from __future__ import annotations

import base64
import json
from typing import Any

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey


def canonical_json(value: dict[str, Any]) -> bytes:
    return json.dumps(value, separators=(",", ":"), sort_keys=True).encode("utf-8")


def verify_activation_payload(payload: dict[str, Any], public_key_b64: str) -> bool:
    signature_b64 = payload.get("signature")
    if not signature_b64:
        return False

    unsigned = dict(payload)
    unsigned.pop("signature", None)

    try:
        signature = base64.urlsafe_b64decode(signature_b64 + "===")
        public_key = Ed25519PublicKey.from_public_bytes(
            base64.b64decode(public_key_b64)
        )
        public_key.verify(signature, canonical_json(unsigned))
        return True
    except (InvalidSignature, ValueError):
        return False

File: licensing/sdk/javascript/ras-license-client.js
export function canonicalJson(value) {
    if (value === null || typeof value !== "object") {
        return JSON.stringify(value);
    }
    if (Array.isArray(value)) {
        return `[${value.map(canonicalJson).join(",")}]`;
    }
    return `{${Object.keys(value).sort().map(key => `${JSON.stringify(key)}:${canonicalJson(value[key])}`).join(",")}}`;
}

function base64UrlToBytes(value) {
    const padded = value.replace(/-/g, "+").replace(/_/g, "/") + "===";
    const raw = atob(padded);
    return Uint8Array.from(raw, char => char.charCodeAt(0));
}

function base64ToBytes(value) {
    const raw = atob(value);
    return Uint8Array.from(raw, char => char.charCodeAt(0));
}

export async function verifyActivationPayload(payload, publicKeyBase64) {
    if (!payload?.signature) return false;

    const unsigned = { ...payload };
    delete unsigned.signature;

    try {
        const key = await crypto.subtle.importKey(
            "raw",
            base64ToBytes(publicKeyBase64),
            { name: "Ed25519" },
            false,
            ["verify"]
        );

        return await crypto.subtle.verify(
            "Ed25519",
            key,
            base64UrlToBytes(payload.signature),
            new TextEncoder().encode(canonicalJson(unsigned))
        );
    } catch {
        return false;
    }
}
